rfcs/008-map-indexes.md

```diff
@@ -0,0 +1,181 @@
+
+Indexes will use the changes feed and be updated via the background tasks queue. If the index only needs a very small update, the update can happen in the request instead of via the background job queue.
+
+There will be new size limitations on keys (10KB) and values (100KB) that are emitted from a map function.
+
+## Requirements Language
+
+[NOTE]: # ( Do not alter the section below. Follow its instructions. )
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+`Sequence`: a 13 byte value formed by combining the current `Incarnation` of the database and the `Versionstamp` of the transaction. Sequences are monotonically increasing even when a database is relocated across FoundationDB clusters. See (RFC002)[LINK TBD]  for a full explanation.
+
+`View Signature`:  A md5 hash of the views, options, view language defined in a design document.
+
+---
+
+## Detailed Description
+
+CouchDB views are used to create secondary indexes for the documents stored in a CouchDB database. An index is defined by creating a map/reduce functions in a design document. This document describes building the map indexes on top of FoundationDB (FDB).
+
+### Data model
+
+A map index is created via a design document, an example is shown below:
+
+```json
+{
+  "_id": "_design/design-doc-id",
+  "_rev": "1-8d361a23b4cb8e213f0868ea3d2742c2",
+  "views": {
+    "map-view": {
+      "map": "function (doc) {\n  emit(doc._id, 1);\n}"
+    }
+  },
+  "language": "javascript"
+}
+```
+
+The view’s map function will be used to generate keys and values which will be stored in FDB as the secondary index. Format for storing the key/values is:
+
+```json
+{<database>, ?VIEWS, <view_signature>, ?VIEW_KVS, <view_id>, ?MAP, <key>, <count>, <_id>} -> <emitted_value>
+```
+
+Where each field is defined as:
+
+* `<database>` is the specific database namespace
+* `?VIEWS` is the standard views namespace.
+* `view_signature` is the design documents `View Signature`
+* `view_id` name of a view defined in the design document
+* `?MAP` is the standard map namespace
+* `key` is the emitted row key from a map function
+* `count` is a value to allow duplicate keys to be emitted for a document
+* `values` is the emitted value from the map function
+
+### Key ordering
+
+FoundationDB orders key by byte value which is not how CouchDB currently orders keys. To maintain the way CouchDB currently does view collation, a type value will need to be prepended to each key so that the correct sort order of null < boolean < numbers < strings < arrays < objects is maintained.
```

Comment:
In addition to the type ordering, there's a consideration about how Numbers are represented when used as keys. FoundationDB has a smart encoding for arbitrary-precision integers, floats, doubles, and even arbitrary-precision decimals, each with it's own type code. CouchDB muddles all that together and sorts them all numerically, regardless of the underlying binary representation.

Given that we're talking about data that did a roundtrip through Javascript, do we just coerce all numbers to be double-precision floating point when used as keys?